### PR TITLE
OCF: Add endpoint information

### PR DESCRIPTION
**Pull request by gabrielschulhof**
@zolkis please have a look at the changes!

### CODE REVIEW

**Comment by gabrielschulhof** [comment]
@poussa @zolkis once this PR goes in I will update the test suite to cover the new API surface.

**Comment by zolkis** [comment]
As discussed on the call, I think we should experiment with  defining a separate dictionary for target or destination containing resource id and endpoint. In this way we could avoid adding a parameter to functions.

File: ocf/client.md [hunk 1]
```diff
@@ -87,6 +88,8 @@ Extends `ResourceId`. Used for creating and registering resources, exposes the p
 
  The `interfaces` property MUST at least contain `"oic.if.baseline"`.
 
+ The `endpoints` property is an array of strings. Each string is a URI consisting of an origin (scheme and authority). It MUST contain at least one endpoint.
```
Comment (nagineni) [inline ocf/client.md]: I think this should array of "ep" objects.  Something like: ` "eps": [{"ep": "coap://[fe80::b1d6]:1111", "pri": 2}, {"ep": "coaps://[fe80::b1d6]:1122"}]`

Comment (nagineni) [reply ocf/client.md]: Here is the spec about Endpoint information:

Endpoint is represented by Endpoint information which consists of two items of key-value pair, "ep" and "pri".

- "ep" shall have as its value a URI with the scheme component indicating Transport Protocol Suite and the authority component indicating the Endpoint Locator

- “pri”- When there are multiple Endpoints, "pri" indicates the priority among them. "pri" shall be represented as a positive integer (e.g. "pri": 1) and the lower the value, the higher the priority. The default "pri" value is 1, i.e. when "pri" is not present, it shall be equivalent to "pri": 1.

Comment (zolkis) [reply ocf/client.md]: This makes sense to me.

**Comment by gabrielschulhof** [comment]
OK, so to keep property names as complete words, let's define as follows:

* a type `Endpoint` with properties
    * `origin` to hold the scheme and authority of the endpoint, and
    * `priority` to hold the priority
* a type named `Destination` which contains properties
    * `deviceId` to hold the device ID,
    * `resourcePath` to hold the resource path, and an optional
    * `endpoint` of type `Endpoint` to hold the endpoint chosesn for the
request.

Thus, a `ResourceId`, `Resource`, and `ClientResource` all qualify as a
`Destination`.

**Comment by zolkis** [comment]
>  `origin` to hold the scheme and authority of the endpoint

Is this an [opaque origin](https://html.spec.whatwg.org/multipage/origin.html#concept-origin-opaque) or a [tuple origin](https://html.spec.whatwg.org/multipage/origin.html#concept-origin-tuple), or should it rather be a [URL](https://url.spec.whatwg.org/#concept-url) and named `url`?

**Comment by gabrielschulhof** [comment]
It's a string. I wouldn't call it "url" because it's not a complete URL.
It's just the origin, and the path has to be concatenated with it in order
for it to be useful.

**Comment by zolkis** [comment]
OK, so it's not a proper origin as per WhatWG standard, and it's not a URL either. On the scheme+authority naming [opinions are split](https://stackoverflow.com/questions/32656791/official-name-for-url-scheme-plus-authority), but I think we can call it `origin`. If someone objects, should offer a better name.

**Comment by gabrielschulhof** [comment]
👍

**Comment by gabrielschulhof** [comment]
@zolkis could you please have another look? I have renamed `ResourceId` to `Destination`, and now all CRUDN methods take a `destination` argument, rather than `Resource` or `ResourceId`.

**Comment by gabrielschulhof** [comment]
Closed in 8e0b3b202774d6ea08c95918ef0508d94740edc5.

**Pull request closed without merging**